clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/deploy/ops/CloneKubernetesAtomicOperation.groovy

```diff
@@ -0,0 +1,109 @@
+ */
+
+package com.netflix.spinnaker.clouddriver.kubernetes.deploy.ops
+
+import com.netflix.frigga.Names
+import com.netflix.spinnaker.clouddriver.data.task.Task
+import com.netflix.spinnaker.clouddriver.data.task.TaskRepository
+import com.netflix.spinnaker.clouddriver.kubernetes.deploy.KubernetesUtil
+import com.netflix.spinnaker.clouddriver.kubernetes.deploy.description.CloneKubernetesAtomicOperationDescription
+import com.netflix.spinnaker.clouddriver.kubernetes.deploy.description.KubernetesContainerDescription
+import com.netflix.spinnaker.clouddriver.kubernetes.deploy.description.KubernetesResourceDescription
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperation
+import io.fabric8.kubernetes.api.model.ReplicationController
+import org.springframework.beans.factory.annotation.Autowired
+
+class CloneKubernetesAtomicOperation implements AtomicOperation<Void> {
+  private static final String BASE_PHASE = "CLONE_SERVER_GROUP"
+
+  @Autowired
+  KubernetesUtil kubernetesUtil
+
+  CloneKubernetesAtomicOperation(CloneKubernetesAtomicOperationDescription description) {
+    this.description = description
+  }
+
+  private static Task getTask() {
+    TaskRepository.threadLocalTask.get()
+  }
+
+  CloneKubernetesAtomicOperationDescription description
+
+/*
+ * curl -X POST -H "Content-Type: application/json" -d  '[ { "cloneServerGroup": { "source": { "serverGroupName": "k8s-test-v000" }, "credentials":  "my-k8s-account" } } ]' localhost:7002/kubernetes/ops
+ * curl -X POST -H "Content-Type: application/json" -d  '[ { "cloneServerGroup": { "stack": "prod", "freeFormDetails": "mdservice", "targetSize": "4", "source": { "serverGroupName": "k8s-test-v000" }, "credentials":  "my-k8s-account" } } ]' localhost:7002/kubernetes/ops
+*/
+  // TODO: should return DeploymentResult object
+  @Override
+  Void operate(List priorOutputs) {
+    CloneKubernetesAtomicOperationDescription newDescription = cloneAndOverrideDescription()
+
+    task.updateStatus BASE_PHASE, "Initializing copy of server group for " +
+      "${description.application}-${description.stack}..."
+
+    new DeployKubernetesAtomicOperation(newDescription).operate(priorOutputs)
+
+    task.updateStatus BASE_PHASE, "Finished copying server group for " +
+      "${description.application}-${description.stack}. "
+  }
+
+  // Made public for testability, not intended to be called directly
+  CloneKubernetesAtomicOperationDescription cloneAndOverrideDescription() {
+    CloneKubernetesAtomicOperationDescription newDescription = description.clone()
+
+    task.updateStatus BASE_PHASE, "Reading ancestor server group $description.source.serverGroupName..."
+
+    ReplicationController ancestorServerGroup = kubernetesUtil.getReplicationController(description.kubernetesCredentials, description.source.serverGroupName)
+
+    if (!ancestorServerGroup) {
+      throw IllegalArgumentException("Source server group $description.source.serverGroupName does not exist")
```

Comment:
Just wondering, why is `IllegalState` better than `IllegalArgument`? It seems like this exception is being thrown in response to a non-existent replication controller provided as an argument to clone.